### PR TITLE
New Function: Search rank data

### PR DESCRIPTION
**Pull request by TheNaman047**
This PR contains code for new function for searching rank data.

### CODE REVIEW

File: functions/src/rankings/index.ts [hunk 1]
```diff
@@ -1,9 +1,65 @@
...
+    if (results.rows[0].count === 0) {
+      return {
+        isError: true,
+        message: 'Please track keyword first',
+      };
+    }
+
+    const listingTableName = getListingTableName(trimmedCountryCode);
+    let searchQuery = `SELECT LZ_LISTINGS.id, LZ_LISTINGS.listing_id, LZ_LISTINGS.scrape_ts, LZ_LISTINGS.search_rank, LZ_LISTINGS.date_str, keywords, T_LISTINGS.country
+                          FROM ${listingTableName} AS LZ_LISTINGS
```
Comment (jeffdh5) [inline functions/src/rankings/index.ts:34]: formatting?

File: functions/src/rankings/index.ts [hunk 2]
```diff
@@ -1,9 +1,65 @@
...
+    }
+    if (startDate && endDate) {
+      searchQuery += ` AND LZ_LISTINGS.DATE_STR BETWEEN '${startDate}' AND '${endDate}'`;
+    }
+
+    const searchResults = await dbPool.query(searchQuery);
+
+    return {
+      isError: false,
+      data: searchResults.rows,
```
Comment (jeffdh5) [inline functions/src/rankings/index.ts:52]: turn it into a mapping:
{listing_id: rows}

Comment (TheNaman047) [reply functions/src/rankings/index.ts:52]: how will this be shown in the frontend?

Or we just return data for one listing id?

File: functions/src/rankings/index.ts (hunk 1 above)
Comment (TheNaman047) [reply functions/src/rankings/index.ts]: Done

File: functions/src/rankings/index.ts [hunk 3]
```diff
@@ -1,9 +1,100 @@
...
+							LZ_LISTINGS.currency,
+							LZ_LISTINGS.rank,
+							LZ_LISTINGS.product_id,
+							LZ_LISTINGS.cleaned_url,
+							keywords,
+							T_LISTINGS.country
+						FROM
+							${listingTableName} AS LZ_LISTINGS
+							JOIN L_TRACKEDLISTINGS AS T_LISTINGS ON LZ_LISTINGS.LISTING_ID = T_LISTINGS.LISTING_ID,
+							UNNEST(T_LISTINGS.TRACKEDKEYWORDS) KEYWORDS
```
Comment (jeffdh5) [inline functions/src/rankings/index.ts:60]: i'm assuming this UNNEST query is going away once we stop using the keywords array and make it 1 row per (user_id, keyword) right?

File: functions/src/rankings/index.ts [hunk 4]
```diff
@@ -1,9 +1,100 @@
...
+						WHERE
+							LOWER(KEYWORDS) like '${trimmedKeyword}'
+							AND T_LISTINGS.USER_ID = '${uid}'`;
+
+    if (listingId) {
+      searchQuery += ` AND T_LISTINGS.LISTING_ID = '${listingId}'`;
+    }
+    if (startDate && endDate) {
+      searchQuery += ` AND LZ_LISTINGS.DATE_STR BETWEEN '${startDate}' AND '${endDate}'`;
```
Comment (jeffdh5) [inline functions/src/rankings/index.ts:69]: Add support for passing just startDate OR just endDate

File: functions/src/rankings/index.ts (hunk 3 above)
Comment (TheNaman047) [reply functions/src/rankings/index.ts:60]: Yes it will go away once that change is done.

File: functions/src/rankings/index.ts (hunk 4 above)
Comment (TheNaman047) [reply functions/src/rankings/index.ts:69]: ok understood, making the change.

File: functions/src/schemas/table.sql [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    price double precision,
+    original_price double precision,
+    discount character varying(16),
+    currency character varying(16),
+    keyword character varying(255) NOT NULL,
+    rank integer NOT NULL,
+    product_id character varying(128),
+    cleaned_url character varying(512),
+    date_str character varying(16),
+    CONSTRAINT l_lazada_id_listing_pkey PRIMARY KEY (keyword, rank)
```
Comment (jeffdh5) [inline functions/src/schemas/table.sql:44]: keyword, rank, scrape_ts

File: functions/src/schemas/table.sql [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    price double precision,
+    original_price double precision,
+    discount character varying(16),
+    currency character varying(16),
+    keyword character varying(255) NOT NULL,
+    rank integer NOT NULL,
+    product_id character varying(128),
+    cleaned_url character varying(512),
+    date_str character varying(16),
+    CONSTRAINT l_lazada_my_listing_pkey PRIMARY KEY (keyword, rank)
```
Comment (jeffdh5) [inline functions/src/schemas/table.sql:66]: update pk

File: functions/src/schemas/table.sql [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    price double precision,
+    original_price double precision,
+    discount character varying(16),
+    currency character varying(16),
+    keyword character varying(255) NOT NULL,
+    rank integer NOT NULL,
+    product_id character varying(128),
+    cleaned_url character varying(512),
+    date_str character varying(16),
+    CONSTRAINT l_lazada_ph_listing_pkey PRIMARY KEY (keyword, rank)
```
Comment (jeffdh5) [inline functions/src/schemas/table.sql:88]: update pk

File: functions/src/schemas/table.sql [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    price double precision,
+    original_price double precision,
+    discount character varying(16),
+    currency character varying(16),
+    keyword character varying(255) NOT NULL,
+    rank integer NOT NULL,
+    product_id character varying(128),
+    cleaned_url character varying(512),
+    date_str character varying(16),
+    CONSTRAINT l_lazada_sg_listing_pkey PRIMARY KEY (keyword, rank)
```
Comment (jeffdh5) [inline functions/src/schemas/table.sql:110]: update pk

File: functions/src/schemas/table.sql [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    price double precision,
+    original_price double precision,
+    discount character varying(16),
+    currency character varying(16),
+    keyword character varying(255) NOT NULL,
+    rank integer NOT NULL,
+    product_id character varying(128),
+    cleaned_url character varying(512),
+    date_str character varying(16),
+    CONSTRAINT l_lazada_vn_listing_pkey PRIMARY KEY (keyword, rank)
```
Comment (jeffdh5) [inline functions/src/schemas/table.sql:132]: update pk

**Pull request merged**